### ISSUE TITLE
Change Proposal: Reorganizing MQTT Topics

### ISSUE THREAD

**Issue by timothybrown** [Aircoookie/WLED#207]
Okay, so I'm currently working on rewriting the HA Discovery code from the ground up. Part of this requires me to basically redo the MQTT topics. Right now it's kind of a mess and none of the topics are ideal for Home Assistant (or other home automation platforms).

The first step is change the topic names and their payload. Here is what I've come up with:

- Power Status Topic: wled/f00b00/status  => *on* / *off*
- Power Command Topic: wled/f00b00/status/set  => *on|true|1* / *off|false|0*
- Color Status Topic: wled/f00b00/color => *R,G,B*
- Color Command Topic: wled/f00b00/color/set => *#HEX*|*R,G,B*
- Brightness Status Topic: wled/f00b00/brightness => *0-255*
- Brightness Command Topic: wled/f00b00/brightness/set => *0-255*
- Effect Status Topic: wled/f00b00/fx => *FX Number*
- Effect Command Topic: wled/f00b00/fx/set => *FX Number*|*FX Name*
- API Response Topic: wled/f00b00/api => *XML API Response*
- API Command Topic: wled/f00b00/api/set => *HTTP API Format*

This puts the topic names in-line with other projects (Espurna, ESPHome, etc.) and makes it easier to integrate into custom code and various home automation platforms without resorting to complex templating or XML decoding (however, because we've retained the API topic, this *can* still be done for advanced users).

Judging by a look through the issue list, it appears most people are using MQTT for HA Discovery support, which means this change will be transparent for those users (i.e., nothing will break for them and it will *just work*). This will only be a breaking change for those who manually configure HA, use other home automation platforms or use MQTT as a transport for custom software. I feel this is a very small percentage of use cases and this change will ultimately make it easier to use for them.

I'm interested to hear from anyone using the MQTT support in a non-HA setting (i.e., with custom code, other automation platforms and so forth). Would these topics be sufficient for your use? Is there anything else you'd like to see?

Comments from others (contributors, users and @Aircoookie) are welcomed as well!

I've currently got the basics of this implemented in my development fork (not yet published) and it's working well, but I wanted to get a discussion going before I go any further with it.

**Comment by timothybrown** [comment]
I should also add that, because of the way I've got this structured, when @Aircoookie finishes support for segments, I'll be able to add segment support into the MQTT code as well. What this means is you could have each segment show up as a separate light in Home Assistant, or you could easily control each segment through MQTT using your custom code. The idea is, each segment would append a number onto the base device topic:
*Device Topic*/0/*status*|*color*|*brightness*|*fx*|*api*
*Device Topic*/0/*status*|*color*|*brightness*|*fx*|*api*/set
*Device Topic*/1/*status*|*color*|*brightness*|*fx*|*api*
*Device Topic*/1/*status*|*color*|*brightness*|*fx*|*api*/set
*Device Topic*/2/*status*|*color*|*brightness*|*fx*|*api*
*Device Topic*/2/*status*|*color*|*brightness*|*fx*|*api*/set
This has a lot of possibilities!

**Comment by brentbrooks70** [comment]
@timothybrown This would ease my implementation of segment control from the backend platform([QIOT Suite](https://www.qnap.com/solution/qiot-suite-platform/en/)) I am using greatly going forward. I think it would actually make the entire system easier to work with via MQTT. 

1. In the power. or now switch, command should there be a 2 (toggle) option? 
    1a. I guess you could always use the _Device Topic/api/set_ with a T2 so maybe not?
2. I know this `might be a little off-topic (maybe needs its own issue) but is there any possibility of exposing the **COOLING** setting via MQTT (or HTTP API or both). I know this is only used by one FX but I have 2 lights that use the **fire_2012** FX as the norm and I really miss not having that setting in WLED. I don't care if it is exposed on the GUI or not but, I would even prefer it to the speed control  for that FX
3. Again, this could be done  _Device Topic/api/set_  FP but exposing palette control may not be bad either
4. My only other thought would be (again probably needs an issue of its own) to change responses from XML to JSON, I'm having difficulty in getting the MQTT messages parsed properly in QIOT Suite as it always expects JSON for some stupid reason.

Sorry if it seems like I'm rambling, sort of thinking (out loud) in writing.

Thanks for your work on this and asking for opinions before just making changes!!

**Comment by Aircoookie** [comment]
Hi,
thank you for ewriting the MQTT integration! I absolutely agree, the topics I chose were quite arbitrary. This is often a problem when implementing something you are not an expert with. I definitely agree with bringing the implementation in line with other projects and easing the integration into HA and similar services, so I am very grateful for your help! Your choice of topics seems to make sense for me, this is a lot more transparent and easy to use than the current system. Maybe we can add speed, intensity and palette, but that could also be done at a later stage.

In fact I haven't been in contact with anyone who used MQTT outside of the home automation use case. HA is by far the most common, but there are users of openHAB and domoticz. Of course I will notify them about the breaking change, there is a high chance that your proposal will benefit the complexity of their implementation as well though!

About segments, the ESP basics are already implemented, just the Javascript UI and a few finishing touches are missing. I like your proposal, structuring the topics like that is very logical!
You can take a look at wled19_json.ino to get an idea how to access them. Basically, you can use the `strip.getSegment(index)` method to access the `Segment` data object, which contains all the important variables of each segment:

```cpp
typedef struct Segment { // 21 bytes
      uint16_t start;
      uint16_t stop; //segment invalid if stop == 0
      uint8_t speed;
      uint8_t intensity;
      uint8_t palette;
      uint8_t mode;
      uint8_t options; //bit pattern: msb first: transitional tbd tbd tbd tbd paused reverse selected
      uint32_t colors[NUM_COLORS];
      //...
      bool isActive()
      {
        return stop > start;
      }
      uint16_t length()
      {
        return stop - start;
      }
    } segment;
```

`strip.getMaxSegments()` will be helpful, too. If you want to support setting segment bounds via MQTT (i'm not sure whether that would make sense), use `setSegment(index, start, stop)`. Let me know if there is anything you need, I'm open to changes to the internal Segment api!

A consideration to make with segment support in general is which segment's values to return via the default APIs (XML and MQTT without segment number in topic). I don't think that returning the first segment would make sense for all users, so I'm planning to make this "return segment" user configurable. It is hardcoded to `0` right now, but you can already use `getReturnedSegmentId()`

I have two small questions:
- What about the group topics (like /wled/all)? HA probably won't use them (because I assume there are better ways to group devices). Do you think there are use cases for them or should we drop them completely? If we keep them, they should of course mirror the device topic names, but only subscribe to the /set topics, not publish anything.

- I'm a little worried that subscribing and publishing to so many topics (especially if we want to support segments) might cause performance issues or even crashing (because of memory overflow by simultaneous requests or similar). We need to make sure that doesn't happen, but for QoS 0 and local MQTT brokers, it could work out.

Thanks once again! Looking forward to hearing from you and let me know if there are any issues!

**Comment by brentbrooks70** [comment]
> What about the group topics (like /wled/all)? HA probably won't use them (because I assume there are better ways to group devices). Do you think there are use cases for them or should we drop them completely? If we keep them, they should of course mirror the device topic names, but only subscribe to the /set topics, not publish anything.

@Aircoookie - IMHO you could drop the ALL, topic, every HA setup (including my own) has grouping abilities or if you really needed groups couldn't you configure a UDP port per group and the UDP sync would handle it for you?

**Comment by timothybrown** [comment]
Yes, I kind of feel that MQTT grouping could be dropped without affecting anyone. As @brentbrooks70 mentioned, we already have the built-in UDP sync functionality that does the same thing, plus most home automation platforms have the ability to group multiple devices into a single “virtual” device.

I’ll give you a good example: My bathroom vanity has three cheap Ai-Light RGBW WiFi bulbs installed (running Espurna). Now, I could use Espurna’s MQTT Group Topic to have them all listen to the same topic and only add one of them to HA, but because they support HA Discovery, it’s easier for me to simply let HA find all three bulbs and create a light group that presents them as a single virtual bulb.

In fact, I’m hard pressed to think of a setup that UDP Sync or Light Groups wouldn’t cover for WLED. (But of course I don’t know every setup, so if anyone can think of one please share!)

**Comment by aidbish** [comment]
Updating the MQTT topics to work correctly will be a big win especially for those of us running HA. Also agree that the group topic can be removed without any if at all detriment as noted above most automation software allows you to create groups.

Happy to be a willing tester of the implementation

**Comment by JAMESBOWLER** [comment]
I argree its not too hard for me to change. I new around here and use openhab mqtt for control.

Looking at the HA docs for mqtt the JSON schema alows the most features to be used. https://www.home-assistant.io/components/light.mqtt/#json-schema

**Comment by timothybrown** [comment]
### @brentbrooks70 

> @timothybrown This would ease my implementation of segment control from the backend platform([QIOT Suite](https://www.qnap.com/solution/qiot-suite-platform/en/)) I am using greatly going forward. I think it would actually make the entire system easier to work with via MQTT.

Yes, that was my thought as well!

> 1. In the power. or now switch, command should there be a 2 (toggle) option?
>    1a. I guess you could always use the _Device Topic/api/set_ with a T2 so maybe not?

Yes, I could easily add a toggle command if it would be useful for you. I had thought about it, but left it out because I couldn’t immediately come up with a use case for it. (Most home automation software will simply track the state by subscribing to the *power* topic, or keep track of the state itself and work in a so called optimistic mode.)

> 2. I know this `might be a little off-topic (maybe needs its own issue) but is there any possibility of exposing the **COOLING** setting via MQTT (or HTTP API or both). I know this is only used by one FX but I have 2 lights that use the **fire_2012** FX as the norm and I really miss not having that setting in WLED. I don't care if it is exposed on the GUI or not but, I would even prefer it to the speed control  for that FX

Please open an issue for this and I’ll look into adding it to the API! @Aircoookie may have some thoughts on it too.

> 3. Again, this could be done  _Device Topic/api/set_  FP but exposing palette control may not be bad either

See my reply to @Aircoookie below.

> 4. My only other thought would be (again probably needs an issue of its own) to change responses from XML to JSON, I'm having difficulty in getting the MQTT messages parsed properly in QIOT Suite as it always expects JSON for some stupid reason.

Yes, I planned on opening an issue about this as well! I’m new as a contributor to the project, so I wanted to get @Aircoookie’s thoughts on why went with XML instead of JSON for responses. XML is a *very* verbose schema compared to JSON. We’ve already got the dependency baggage of the JSON library, so why not go all the way and have responses in JSON? It would use less memory, bandwidth and flash, plus it’s easier to encode/decode when writing custom software to control your devices. 

I’ll open an issue about this and we can discuss it there.

> Sorry if it seems like I'm rambling, sort of thinking (out loud) in writing.

You’re not rambling at all. Personally, I like people thinking out loud like this in a development environment; it’s one of the things that make open source and communities like GitHub so great!

> Thanks for your work on this and asking for opinions before just making changes!!

And thank *you* for the feedback, keep it coming!

### @Aircoookie 
> Hi,
> thank you for ewriting the MQTT integration! I absolutely agree, the topics I chose were quite arbitrary. This is often a problem when implementing something you are not an expert with.

No worries, at the very least you got basic MQTT support working, which is a lot better than nothing!

> I definitely agree with bringing the implementation in line with other projects and easing the integration into HA and similar services, so I am very grateful for your help! Your choice of topics seems to make sense for me, this is a lot more transparent and easy to use than the current system. Maybe we can add speed, intensity and palette, but that could also be done at a later stage.

Yes, I was thinking about adding topics for speed, intensity and palette too, but I wasn’t sure how useful they would be. The Home Assistant MQTT Light component doesn’t have corresponding controls for those options (other home automation platforms appear to be similar) so they would only be useful in HA if:

1. You’re using a custom version of the MQTT Light component that you’ve added these controls to.
2. You add input_* control components (like input_select, input_boolean or input_number) and combined with automation rules or an AppDaemon script to map them to the relevant topics for each light. 
3. You could also map speed and intensity to the *white_value_* and *color_temperature_* controls, since we don’t use them. (You would still have to add another component (like *input_select*) and write an automation rule to handle the palette for each light.)

However, in all of these cases you can do the same thing by just using the API topic. You’re already implementing custom scripts or automation rules, at which point it’s trivial to just decode and encode for the API topic.

That said, it’s trivial to add them, so if it will make it easier for someone to integrate into their platform or custom code I can add them to the list!

> In fact I haven't been in contact with anyone who used MQTT outside of the home automation use case. HA is by far the most common, but there are users of openHAB and domoticz. Of course I will notify them about the breaking change, there is a high chance that your proposal will benefit the complexity of their implementation as well though!

Thanks!

> About segments, the ESP basics are already implemented, just the Javascript UI and a few finishing touches are missing. I like your proposal, structuring the topics like that is very logical!
> You can take a look at wled19_json.ino to get an idea how to access them. Basically, you can use the `strip.getSegment(index)` method to access the `Segment` data object, which contains all the important variables of each segment:
> …
> `strip.getMaxSegments()` will be helpful, too. If you want to support setting segment bounds via MQTT (i'm not sure whether that would make sense), use `strip.setSegment(index, start, stop)`. Let me know if there is anything you need, I'm open to changes to the internal Segment api!

Okay, cool. I’ll start diving into this over the weekend and get a first pass implemented. I may be in contact with some questions as I go!

> A consideration to make with segment support in general is which segment's values to return via the default APIs (XML and MQTT without segment number in topic). I don't think that returning the first segment would make sense for all users, so I'm planning to make this "return segment" user configurable. It is hardcoded to `0` right now, but you can already use `strip.getReturnedSegmentId()`

Yes, I was thinking about this earlier. I think the easiest way to handle this on the MQTT side would be to *always* include the segment number in the topic path. So, even if the user doesn’t use segments their topic would be <Device Topic>/0/<Command> because 0 is the default segment consisting of the entire string of LEDs.

> I have two small questions:
> 
> * What about the group topics (like /wled/all)? HA probably won't use them (because I assume there are better ways to group devices). Do you think there are use cases for them or should we drop them completely? If we keep them, they should of course mirror the device topic names, but only subscribe to the /set topics, not publish anything.

See my earlier reply for more on this.

> * I'm a little worried that subscribing and publishing to so many topics (especially if we want to support segments) might cause performance issues or even crashing (because of memory overflow by simultaneous requests or similar). We need to make sure that doesn't happen, but for QoS 0 and local MQTT brokers, it could work out.

I don’t think this will be a big issue, but I planned to do some performance profiling and see how far I can push it. The only place we may run into an issue is if the user has *a lot* of segments. In those cases they might have to use an ESP32.

The alternative is to only have a single topic per segment which would it be a JSON input and output. Home Assistant’s MQTT Light component supports three types of schemes: default, JSON and template. Right now I’m designing it for the default schema, where the various controls have their own topics. We *could* switch to the JSON schema, which basically expects a status topic that contains a JSON dictionary, likewise it sends a JSON dictionary to the command topic. Here’s an example:
```
{
  "brightness": 255,
  "color_temp": 155,
  "color": {
    "r": 255,
    "g": 180,
    "b": 200,
    "x": 0.406,
    "y": 0.301,
    "h": 344.0,
    "s": 29.412
  },
  "effect": "colorloop",
  "state": "ON",
  "transition": 2,
  "white_value": 150
}
```

> 
> Thanks once again! Looking forward to hearing from you and let me know if there are any issues!

And thank *you* for this great project! I’m happy to be able to contribute something.

**Comment by timothybrown** [comment]
> I argree its not too hard for me to change. I new around here and use openhab mqtt for control.
> 
> Looking at the HA docs for mqtt the JSON schema alows the most features to be used. https://www.home-assistant.io/components/light.mqtt/#json-schema

Yes, compared to the default schema, the JSON schema gives you a flashing/strobing control and a control to set the transition time between colors. The former isn’t really applicable to WLED and I’m not sure how useful the latter would be.

If you guys would *prefer* the JSON schema (as described at the end my last post) I could absolutely implement it instead.

**Comment by aidbish** [comment]
i think the using the json schema would give greater flexibility going forward.

I like the idea of the speed and intensity of an effect being exposed via MQTT as this would for me enable me to set effects with certain speeds in nodered be able to use a zigbee remote to  use

**Comment by brentbrooks70** [comment]
@timothybrown 

> I don’t think this will be a big issue, but I planned to do some performance profiling and see how far I can push it. The only place we may run into an issue is if the user has a lot of segments. In those cases they might have to use an ESP32.

When you are ready to performance test let me know, I can script some pretty rapid fire calls via QIOT to see how fast it can accept the calls and just how quickly it chews up memory for both platforms. 

Very excited about these changes, WLED just keeps getting better!!!

**Comment by timothybrown** [comment]
> i think the using the json schema would give greater flexibility going forward.

Yes, I tend to agree, however...

> I like the idea of the speed and intensity of an effect being exposed via MQTT as this would for me enable me to set effects with certain speeds in nodered be able to use a zigbee remote to use

Keep in mind that you’d already be able to do that by just passing the correct command (for speed, intensity or palette) to the API topic. The real question is if it’s worth it to expose those commands as their own unique topics. It seems like in your case either would work, though exposing them as topics would make it simpler for you to implement.

What I’m kind of thinking about right now is maybe going with the topics I outlined in my original post, but changing the API topic to accept a JSON input and provide a JSON output. The JSON schema wouldn’t necessarily conform to the default HA JSON schema, in order to save memory, but *would* still expose all the available commands. This would be fine as HA would be using the default schema for all the basic functions, the JSON topic would just be there for advanced users or people like you using other platforms.

Hopefully that makes sense, I’m just thinking out loud here.

> When you are ready to performance test let me know, I can script some pretty rapid fire calls via QIOT to see how fast it can accept the calls and just how quickly it chews up memory for both platforms.
> 
> Very excited about these changes, WLED just keeps getting better!!!

Awesome, this will be very helpful. I’ve only got two custom lamps that have 70 LEDs in them (that I mainly use with the fire effect), plus an additional strip of 100 on a wall as bias lighting (I’m slowly expanding that one to eventually encircle the entire room). So testing with larger arrays and larger numbers of smaller arrays will be great. (The latter is important because network traffic can seriously slow things down, especially when you’ve got a bunch of WiFi devices all rapidly sending packets).

**Comment by Aircoookie** [comment]
@brentbrooks70 
See #208 !

@timothybrown 

>  Yes, I kind of feel that MQTT grouping could be dropped without affecting anyone.

Seems like we reached a consensus. Let's drop it! If the user only wants to connect some of the lights to the broker for some reason, they can still use UDP sync!

> You could also map speed and intensity to the _white_value_...

That is a possibility, although the custom implementation seems more straighforward. Having white value and color temperature do something completely different would confuse new users and it still doesn't cover palettes. That said, WLED does offer RGBW support for WS2813 strips. The bool `useRGBW` is true if the user enabled it. The white value is either `col[3]` or for segments, the 8 most significant bytes of the color uint32_t. Maybe it is possible to add that topic when RGBW is enabled, not sure on that one!

> So, even if the user doesn’t use segments their topic would be Device Topic/0/Command because 0 is the default segment consisting of the entire string of LEDs.

Agreed! This is a lot easier both for implementation and for the user.

On the topic of JSON vs. topic-per-property scheme types: I think you should implement this however you see fit! If there are any issues down the line we can still switch to the other type.

> My thought is that the JSON API is new and maybe his plan is to eventually switch over to it.

This exactly! MQTT was implemented before the JSON API and I have not yet connected them. Since JSON has become easier to deal with in just about any modern environment, I think this reorganization is a good opportunity to provide a JSON state API topic for MQTT in addition to or instead of the XML one. This shouldn't be too much work because we can just hook to the provided functions `serializeState(JsonObject)` and `deserializeState(JsonObject)`. Heres the [API doc](https://github.com/Aircoookie/WLED/wiki/JSON-API) 

> Okay, cool. I’ll start diving into this over the weekend and get a first pass implemented. I may be in contact with some questions as I go!

Awesome, thanks!

**Comment by timothybrown** [comment]
What do you guys think about adding some general status topics to the list? These would be published under the root *device topic* only and not duplicated for each segment. Here's an example of what Espurna publishes:

```
Bathroom-Vanity-Left/app ESPURNA
Bathroom-Vanity-Left/version 1.13.5
Bathroom-Vanity-Left/board AITHINKER_AI_LIGHT
Bathroom-Vanity-Left/host Bathroom-Vanity-Left
Bathroom-Vanity-Left/ssid ********
Bathroom-Vanity-Left/ip 10.***.***.***
Bathroom-Vanity-Left/Mac **:**:**:**:**:**
Bathroom-Vanity-Left/rssi -52
Bathroom-Vanity-Left/uptime 233119
Bathroom-Vanity-Left/datetime 2019-08-24 21:15:12
Bathroom-Vanity-Left/freeheap 21944
Bathroom-Vanity-Left/vcc 3246
Bathroom-Vanity-Left/status 1
Bathroom-Vanity-Left/loadavg 1
```

**Comment by aidbish** [comment]
Yep i like, you could make sensors in HA for things like, fw version, ip address which saves having to load the gui and could be on a glance card or more info card

**Comment by timothybrown** [comment]
@aidbish Yup! I was just reading through some HA documents and came across another way that might be viable as well. Instead of having all those individual status topics, we could just have a single one that publishes a JSON payload, like so:
*Device Topic*/*sysinfo*
```
{
  “Firmware”: “WLED *xx.yy.zz*”,
  “VCC”: “33000”,
  “RSSI”: “-50”,
  “MAC”: “00:DE:AD:BE:EF:00”,
  “Hostname”: “wled.local”,
  “IP”: “127.0.0.1”,
  “SSID”: “Generic Router”,
  “Load”: “1.0”,
  “Memory”: “50000”,
  “Uptime”: “300”
}
```

Now, the advantage here is that HA will actually display this information in the UI without requiring you to add additional sensors (the attributes will be at the very bottom of the card). You can also access the data from scripts or automations via simple templating.

To get this working, all we need to do is include an entry for `json_attributes_topic` that points to *Device Topic*/*sysinfo* in our HA Discovery config payload.

This seems like an overall much cleaner and more useful way to do it. What do you guys think?

**Comment by aidbish** [comment]
That would work, 
i am happy with either way, as long as we get the correct status of the strip and can manipulate its options.

i think once done, we post on the HA about this, At the moment i am finishing off some outdoor glass rock lights with 16pixel circular strips, (using recent build of WLED) so happy to do some show and tell when mqtt is working on above to promote WLED.

**Comment by timothybrown** [comment]
Okay, here’s my revised list. Did I forget anything?

### Segment Topics
- Power Status Topic
*Device Topic*/*Segment*/switch  => *on*|*off*
- Power Command Topic
*Device Topic*/*Segment*/switch/set  <= *on*|*true*|*1*|*off*|*false*|*0*
- Color Status Topic
*Device Topic*/*Segment*/color => *R,G,B*
- Color Command Topic
*Device Topic*/*Segment*/color/set <= *#HEX*|*R,G,B*
- Brightness Status Topic
*Device Topic*/*Segment*/brightness => *0-255*
- Brightness Command Topic
*Device Topic*/*Segment*/brightness/set <= *0-255*
- White Value Status Topic (*For RGBW LEDs)
*Device Topic*/*Segment*/white => *0-255*
- White Value Command Topic (*For RGBW LEDs)
*Device Topic*/*Segment*/white/set <= *0-255*
- Effect Status Topic
*Device Topic*/*Segment*/fx => *FX Number*
- Effect Command Topic
*Device Topic*/*Segment*/fx/set <= *FX Number*|*FX Name*
- Palette Status Topic
*Device Topic*/*Segment*/palette => *Palette Name*
- Palette Command Topic
*Device Topic*/*Segment*/palette/set <= *Palette Name*
- Intensity Status Topic
*Device Topic*/*Segment*/intensity => *0-255*
- Intensity Command Topic
*Device Topic*/*Segment*/intensity/set <= *0-255*
- Speed Status Topic
*Device Topic*/*Segment*/speed => *0-255*
- Speed Command Topic
*Device Topic*/*Segment*/speed/set <= *0-255*

### Non-Segment Topics
- API Response
*Device Topic*/api => *JSON API Response*
- API Command
*Device Topic*/api/set <= *JSON API Format*
- Status Topics
*Device Topic*/status => *online*|*offline*
*Device Topic*/info => *JSON Attributes Response*
```
{
  “Firmware”: “WLED xx.yy.zz”,
  “VCC”: “3300”,
  “RSSI”: “-50”,
  “MAC”: “00:DE:AD:BE:EF:00”,
  “Hostname”: “wled.local”,
  “IP”: “127.0.0.1”,
  “SSID”: “Generic Router”,
  “LoadAvg”: “1.0”,
  “FreeMem”: “50000”,
  “Uptime”: “300”
}
```

**Comment by aidbish** [comment]
Looks good from me. 
A good outcome.

**Comment by brentbrooks70** [comment]
@timothybrown That looks very good, the JSON attribute response makes way more sense than the individual topics to me.

**Comment by timothybrown** [comment]
Okay, cool! So that’s what I’m going with then. I should have a test branch available sometime this week. This has taken a bit more time than I initially anticipated, mainly because I had to fully get back into the mindset of programming in C/CPP. (I didn’t realize just how much I’d started taking for granted simple Python functionality like string and type handling!)

**Comment by brentbrooks70** [comment]
@timothybrown  LOL, Strings??? No such thing in C! 
Let me know when you have something I can stress test. I've got both ESP platforms set with 144LEDs (144/m)  as well as both platforms with 300 LEDs (60/m), in a test environment, and an 8266 with 700 LEDs (60/m) ready for testing (this last set is in my daughters room so I can only test during weekdays while she is at school). I have also designed a Node-Red flow that will generate up to 100 messages per sec. Will start with something much slower initially but would like to push it until I kill it, Should I have the board reply to each message with a freeheap (not sure how that would impact memory)  or maybe poll the HTTP freeheap at certain intervals to get an idea of memory consumption? Not sure which method would cause less of an impact on speed or memory use....thoughts?

**Comment by Aircoookie** [comment]
@timothybrown 
Awesome! While looking at the topic list, I forgot something though. Although by far not used by all effects, there is a secondary (background) and even tertiary color. We can either add topics for those or just use the JSON API. Totally agree, when Python is like building a computer, Cpp is like designing the CPU!

@brentbrooks70 Awesome setup! I bet your daughter loves her room! If you want to stress test for X amount of LEDs by the way, you don't need to have them physically connected. Just selecting the amount in LED settings is enough. I've just pushed an update to master that fixes a critical issue with the JSON api (HTTP POST requests had no effect). I plan to release it as 0.8.5 soon. If you'd like to, stresstest it! About freeheap, the new `/info` topic will provide it via MQTT. Or just poll the HTTP as you said. I've found the ESPs to be quite resilient to about 50 requests per second. Above that they are handled too slowly and heap crash will occur. Also another developer has found an apparent issue in the TCP library that can cause memory leaks if there are multiple IPs sending requests to the ESP simultaneously (https://github.com/me-no-dev/ESPAsyncWebServer/issues/528). Would be interesting to know how much of an issue this is in practice!

**Comment by timothybrown** [comment]
> @timothybrown
> Awesome! While looking at the topic list, I forgot something though. Although by far not used by all effects, there is a secondary (background) and even tertiary color. We can either add topics for those or just use the JSON API. Totally agree, when Python is like building a computer, Cpp is like designing the CPU!

Huh, yeah, that's a good point about the secondary/tertiary colors... The first thing that jumps to my mind is, what if we allow inputting that via the */color* topic? We could *optionally* accept a JSON list like so: `{[[255,255,255], [0,0,0], [127,127,127]]}` (Primary, Secondary, Tertiary)

The color status topic would report back in the same way, but only if the command is given in this JSON format.

Normally, I'm not keen on having a command channel behave in more than one way (it tends to greatly complicate end user implementation), but this one might not be too bad. I'll need to think on this one a bit more.

**Comment by brentbrooks70** [comment]
> If you want to stress test for X amount of LEDs by the way, you don't need to have them physically connected. Just selecting the amount in LED settings is enough

@Aircoookie Without having the LEDs attached how would you visibly know if you are impacting the update speed of each "frame" or causing "hickups"? 

> I bet your daughter loves her room!

@Aircoookie  Even more so thanks to you and WLED!!

> Normally, I'm not keen on having a command channel behave in more than one way (it tends to greatly complicate end user implementation), but this one might not be too bad. I'll need to think on it a bit more.

@timothybrown I agree, multiple behaviors on the same command could become confusing; if required heavy documentation might be the only saving grace.

**Comment by timothybrown** [comment]
@Aircoookie I know this is off topic, but are there any plans on moving configuration options over from "EEPROM" storage to SPIFFs? We're starting to run up against constraints and the implementation is becoming exponentially more complex as we add to it. Moving configuration options over to an INI file stored in SPIFFs would greatly simplify things I think.

**Comment by timothybrown** [comment]
Here's the configuration JSON we'll be sending to Home Assistant.

```
{
  "name": "WLED Light (Segment)", # Settings->User Interface->Server Description + "(" + Segment Number + ")"
  "platform":"mqtt",
  "state_topic":"Device Topic/Segment/switch",
  "command_topic":"Device Topic/Segment/switch/set",
  "payload_on":"on",
  "payload_off":"off",
  "availability_topic":"Device Topic/status",
  "payload_available":"online",
  "payload_not_available":"offline",
  "brightness_state_topic":"Device Topic/Segment/brightness",
  "brightness_command_topic":"Device Topic/Segment/brightness/set",
  "rgb_state_topic":"Device Topic/Segment/color",
  "rgb_command_topic":"Device Topic/Segment/color/set",
  "white_value_state_topic":"Device Topic/Segment/white",
  "white_value_command_topic":"Device Topic/Segment/white/set",
  "json_attributes_topic":"Device Topic/info",
  "uniq_id":"WLED-BEEF00_0", # "WLED-" + Last Six of MAC + "_" + Segment Number
  "device":{
    "identifiers":["WLED-BEEF00"], # "WLED-" + Last Six of MAC
    "name":"WLED Light", # Settings->User Interface->Server Description
    "manufacturer":"WLED",
    "model":"Board Name",
    "sw_version":"WLED xx.yy.zz"
    }
  }
```

**Comment by timothybrown** [comment]
Implementation Detail/Note to Future Self: Add a hook to the segment API so that when a user *deletes* a segment we call a function that deletes it from HA as well (by sending an empty configuration payload to the Home Assistant discovery topic).

**Comment by aidbish** [comment]
no need to apologise. as a non developer i like trying to understand the process.

**Comment by JAMESBOWLER** [comment]
Timothy you need to understand somthing that took me years to comprehend. Every one of you contributions and comments on here have helped someone understand what you are doing and what you were thinking at the time. Without these insites, questions and examples the project may lead off in a diferent direction.

**Comment by Aircoookie** [comment]
> I know this is off topic, but are there any plans on moving configuration options over from "EEPROM" storage to SPIFFs? We're starting to run up against constraints and the implementation is becoming exponentially more complex as we add to it. Moving configuration options over to an INI file stored in SPIFFs would greatly simplify things I think.
The reason I bring this up is because I'd like to add another option on the MQTT page to enter the Home Assistant discovery base topic. (If the user leaves it blank we'll disable HA Discovery.)

WLED has SPIFFS support, it can be enabled via headers in wled00.ino. I considered switching to an SPIFFS configuration file and will likely do soon-ish. The main problem would the high memory usage of using descriptive JSON keys. However I believe it would drastically increase transparency and code legibility. Do you know whether it's possible to have F() (or similar, store in flash) strings as JSON keys with ArduinoJSON 6? In that case, i'd highly prefer to switch to a config file in SPIFFS. Another issue is cleartext passwords, I'd rather not store them in an accessable file (at least not the home WiFi password). Maybe not storing them at all explicitely and making use of `WiFi.persistant()` would work, so we could get rid of the current EEPROM implementation entirely.

Also I agree with the others, not writing down one's thoughts or alternatives is kind of like writing code without comments. So I highly appreciate that you do!

**Comment by prankousky** [comment]
I have tried using WLED with Home Assistant, but was not able to get satisfying results, so this thread really got my hopes up.

Are any of you successfully using the current version of WLED with Home Assistant? If so, would you please share your configuration? The template from the Wiki is great, but I am missing options to change palettes.

**Comment by timothybrown** [comment]
@prankousky The changes being discussed in this thread have *not* yet been merged into WLED. In the current version the very basics work (setting power, brightness and FX), but the status reporting from WLED to HA is broken. I’m sure you could come up with a passable config using the XML API topic and templating, but it would ultimately be a waste of time, as it will break when my update gets merged in. I’m hoping to have a beta version of the new MQTT/HA code up on my fork of the repo this weekend, so just hang in there a bit longer, it’ll be worth the wait!

As for palettes, unfortunately Home Assistant does *not* provide support for palettes in the MQTT light component. That component only provides controls for color, brightness, white level, color temperature and effects. That said, I am exposing the palette (and other) options as topics, so it *would* be possible to implement palette support via something like the *input_list* component and either a set of automation rules, a script or AppDaemon.

As an aside I should mention that once I get the new MQTT/HA code stable and released I will start looking into making a custom version of the MQTT Light component that exposes these additional options (palettes, presets, secondary/tertiary colors) that people can install if they want. I don’t think it will be that hard, but it’ll most likely be a couple of months before you see it.

**Comment by timothybrown** [comment]
Okay, so I’ve again updated the list. I decided to give the primary/secondary/tertiary colors their own topics under the color hierarchy. Personally, I feel a better way to handle this would be by adding an option to the API to let the user setup custom palettes and then just select said palette. (However, that’s a discussion for a later time.)

### Segment Topics
- Power
*Device Topic*/*Segment*/switch  => *on*|*off*
*Device Topic*/*Segment*/switch/set  <= *on*|*true*|*1*|*off*|*false*|*0*
- Color
*Device Topic*/*Segment*/color/primary => *R,G,B*
*Device Topic*/*Segment*/color/primary/set <= *#HEX*|*R,G,B*
*Device Topic*/*Segment*/color/secondary => *R,G,B*
*Device Topic*/*Segment*/color/secondary/set <= *#HEX*|*R,G,B*
*Device Topic*/*Segment*/color/tertiary => *R,G,B*
*Device Topic*/*Segment*/color/tertiary/set <= *#HEX*|*R,G,B*
- Brightness
*Device Topic*/*Segment*/brightness => *0-255*
*Device Topic*/*Segment*/brightness/set <= *0-255*
- White Value (*For RGBW LEDs)
*Device Topic*/*Segment*/white => *0-255*
*Device Topic*/*Segment*/white/set <= *0-255*
- Effects
*Device Topic*/*Segment*/fx => *FX Number*
*Device Topic*/*Segment*/fx/set <= *FX Number*|*FX Name*
- Palette
*Device Topic*/*Segment*/palette => *Palette Name*
*Device Topic*/*Segment*/palette/set <= *Palette Name*
- Intensity
*Device Topic*/*Segment*/intensity => *0-255*
*Device Topic*/*Segment*/intensity/set <= *0-255*
- Speed
*Device Topic*/*Segment*/speed => *0-255*
*Device Topic*/*Segment*/speed/set <= *0-255*

### Non-Segment Topics
- API
*Device Topic*/api => *JSON API Response*
*Device Topic*/api/set <= *JSON API Format*
- Status
*Device Topic*/status => *online*|*offline*
*Device Topic*/info => *JSON Attributes Response*
```
{
  “Firmware”: “WLED xx.yy.zz”,
  “VCC”: “3300”,
  “RSSI”: “-50”,
  “MAC”: “00:DE:AD:BE:EF:00”,
  “Hostname”: “wled.local”,
  “IP”: “127.0.0.1”,
  “SSID”: “Generic Router”,
  “LoadAvg”: “1.0”,
  “FreeMem”: “50000”,
  “Uptime”: “300”
}
```

**Comment by Aircoookie** [comment]
@timothybrown Awesome, having separate topic for them might not be a bad idea! I would however propose renaming them to `_Device Topic_/_Segment_/color/0` and so on, using numerals. This would allow us to extend the number of colors at some point (which I won't do since for that palettes are the correct tool, the only reason I haven't switched to palette-based color handling exclusively is the white-channel support). But this syntax would allow for supporting custom palettes over the color topic as well. Custom palettes are something I am definitely planning, but I'm not yet sure on merging the concepts of colors and palettes. Let's keep that discussion for another time though!

Another point I wanted to bring up is that I noticed some users solely use the MQTT implementation to send [HTTP API](https://github.com/Aircoookie/WLED/wiki/HTTP-request-API) requests aka something like `T=2`. It could be benefitial for the `/api/set` topic to still accept that syntax alongside JSON (by falling back to it if JSON parsing fails, or just checking for the `{` character).
You could use this code from the current implementation to call the HTTP API handler:
```cpp
String apiReq = "win&";
apiReq += (char*)payload;
handleSet(nullptr, apiReq);
```

Thanks once again for your hard work! I really like the new topic structure!

**Comment by sujitrp** [comment]
With current 0.8.3 version MQTT  integrated the WLED with OpenHab 

![image](https://user-images.githubusercontent.com/7476632/64722050-4a45c400-d4eb-11e9-9027-b7f6d61e0209.png)


https://github.com/Aircoookie/WLED/issues/35#issuecomment-530479211

**Comment by timothybrown** [comment]
Hey guys, sorry for the delay! Food poising sucks.

I’m back on this today and will be submitting a pull request soon!

**Comment by aidbish** [comment]
Mate, no need to be sorry. Hope your feeling better though? food poisoning isn't fun

**Comment by flashy02** [comment]
@timothybrown : I have found this isue and i am kind happy to read about the planed MQTT refactoring. Is there a timeline when it will be available? At the moment I'm playing with wled an I would wait for integration in OpenHAB until the new MQTT interface is available.

**Comment by EvasionOfTruth** [comment]
I would love for this to be implemented.

**Comment by bkpsu** [comment]
@timothybrown - great work with this so far, excited to see it implemented in WLED, as I'm a new convert/user and LOVE the work @Aircoookie has done with this project (all the different effects, interfaces, just awesome stuff!). I've already got WLED installed on one of my Kube LED controllers, and was getting ready to bring it into my openHAB setup, when I learned about your rewrite of the MQTT interface,

I do have one "dissenting" comment about removing the group topic. While it is true that most of the home automation platforms have a way to group devices, without a group topic, the OH platform needs to send individual messages to each device in the group. With an increasing # of devices (I have a LED strip in every window of my house, configured to synchronize effects for the holidays), the message stack causes delays, and results in a "cascading" fx change. With a group topic, all devices instantly synchronize to the one message meant for all of them. That is the benefit of a device-side group topic, IMO.

Of course, there is a workaround, to have all devices subscribe to the same \set topic, but then you lose ability to control each device individually.

Just my $0.02/suggestion. Keep up the great work!

**Comment by bagobones** [comment]
Hi All, I just discovered WLED, I have been banging away on some custom code based on https://github.com/thehookup/Holiday_LEDs_2.0/blob/master/Holiday_Lights_2.0.yaml which has segment support.

There might be some value in having WLED present one topic but still have segments.. IN the hookup code effects are played in all zones with the same settings but having zones changes the way they appear, IE lights bounce IN each zone or fire appears to burn from the ends of zones.. It is easy to control but looks more custom to your house or installation.

Other than that I would still be super happy with segments each appearing as a discovered light and not bothering with grouping.

I think the main value to having group topics is syncing things and keeping traffic down, if that is the case a group topic for ALL WLEDs (if you are running several instances) and ALL segments in a multi segment unit would be useful for some. Not completely necessary however.

**Comment by TheFuzz4** [comment]
@bagobones that is exactly what I'm looking to do once I get my strips in.  I'd also like to be able to control multiple zones from the various GPIO pins of the NodeMCU.  Thank you @Aircoookie and @timothybrown for your continued hard work.  Looking forward to installing these this year.

**Comment by miesjel12** [comment]
Is there an update on when this will be released in an update? I have WLED installed but not yet connected to Home Assistant because of this. Should I wait or do you advice to just implement it and change it when the update arrives (since this will be a breaking change if I update the WLED controller)?

Good work so far, looks promising!

**Comment by jhuizingh** [comment]
I'm entering the conversation a little late in the game, so if this change is already in motion I think it's a step in the right direction!

It does seem kind of specific to HA.  I use openHAB, and something that would make MQTT as easy as possible to implement is for each different item that can be set to have a unique topic. The reason for this is that in openHAB you define and item, then you can set inbound and outbound mqtt topics in the item metadata. If an item does not have its own topic, then it would have to be parsed from a message that has the info for multiple values. For example:

```
Switch LEDStrip_Power "LED Lights" (gLEDStrip_WLED_Power,gOutdoorLights) ["Lighting"] 
    {mqtt=">[mosquitto:wled/test:command:*:default],
    <[mosquitto:wled/test/v:state:XPATH(concat(substring('OFF',1,number(not(not(//ac/text()[.='0'])))*3),substring('ON',1,number(not(//ac/text()[.='0']))*2)))]"} 
```
For example, the `Power` can output directly to the device topic since it is unique for that purpose. However, from my limited experimenting it looks like that info is only possible in the <ac> tag on the /v incoming topic. There are a number of different ways to transform input, but they all have some pitfalls. So I had a few options:

1. Create an XPATH transform. That's what I did above. It's convoluted at best but converts 0 to OFF and any nonzero number to ON. The pro is that it's all inline.
2. Run the /v data through a Javascript transformation function or maybe an XSLT transformation. Ideally it would be reusable, but as-is at the point of the transformation function we don't know what item/topic the xml is for, so I'd have to create a different transformation file for each discrete piece of WLED data that I wanted in my system ( wled_ac.js, wled_cl.js wled_cs.js, wled_fx.js etc..)

Both of those are tricky or don't scale well to many items.

Sending would be similarly interesting. Probably the best way would be to create a rule for when the majority of the items change, and then pass the proper api format string to /api based on which item has changed (i.e. LEDStrip_Effect -> send /api: FX=value).  Not ideal because there would have to be a mapping of item name to parameter string that needs to be sent to the topic. My ideal mqtt topic format would probably match with the HTTP Request Api and be something like this:

[mqttDeviceTopic]/A
[mqttDeviceTopic]/A/set
[mqttDeviceTopic]/T
[mqttDeviceTopic]/T/set
[mqttDeviceTopic]/FX
[mqttDeviceTopic]/FX/set
[mqttDeviceTopic]/SX
[mqttDeviceTopic]/SX/set
etc...  

It could also use any string such as the ones used in the JSON api instead for the property names. I'm not too concerned which ones are used. I figure code-wise it may be easiest to use what's already defined for one of the other apis.

With the above proposed MQTT topics, you could subscribe to

`[mqttDeviceTopic]/#` which would match all of the above.  Then it could be parsed out using regular expressions.
 
The regular expression  `^[mqttDeviceTopic]\/(\w+)(?:\/(\w+))?$` would match any of the above examples. $1=FX . $2=set . when the topic that received the publish is `[mqttDeviceTopic]/FX/set`.
See examples here:  regexr.com/4noei

After a match, you could use the value of $1 to figure out which property to update, similar to how the HTTP Binding does it (I assume). 


`^[mqttDeviceTopic]\/(\d+)\/(\w+)(?:\/(\w+))?$`
See examples here: regexr.com/4nof1
If this regex was run first, it would catch any segment items because of the \d+ argument, then the previous regex would catch all "top level" topics.

Implementing this would allow openHAB items to be defined more like this:


```
Switch LEDStrip_Power "LED Lights" (gLEDStrip_WLED_Power,gOutdoorLights) ["Lighting"] 
    {mqtt=">[mosquitto:wled/FX/set:command:*:default],
    <[mosquitto:wled/FX:state:JS(FXStringNumberMap)]"} 
```
...Much simpler and more reusable.

Does this look worthwhile at all?

I'm pretty rusty in my C code, otherwise I would try implementing this myself.

**Comment by Aircoookie** [comment]
@jhuizingh Yes, I believe your proposal is pretty much what we are already going for!
We need to make sure that the new topics are useful for as many users as possible!

@timothybrown the only difference I see to your proposed structure at first glance is the absense of the segment ID. I don't believe this is an issue since any application that does not want to support segments can just use `[mqttDeviceTopic]/0` as root topic.

**Comment by jhuizingh** [comment]
@Aircoookie my schema actually attempted to support @timothybrown's as much as possible. It can support the segment id stuff too. The second regex above: 
```
^[mqttDeviceTopic]\/(\d+)\/(\w+)(?:\/(\w+))?$
```

Handles the segment id through the `(\d+)` portion as shown at [regexr.com/4nof1](regexr.com/4nof1)

**Comment by bagobones** [comment]
While I use HA, I think making the topics as universally compatible is a good idea. Since HA supports auto discovery how the topics work on HA wont matter as much to average HA users anyway.

**Comment by Aircoookie** [comment]
Just implemented a tiny part of this in latest master.
`[mqttDeviceTopic]/status` will now be "online" with an automated last will setting it to "offline" should the device lose connection!

**Comment by nightcat91** [comment]
Really loving the wled project and looking forward to the rewrite of the mqtt part.
So far I have used auto discovery from HomeAssistant to add the wled lights to my config.
Its working well so far (with the known limitations).

One thing I wanted to bring up is that currently HA cannot get the state of the light (is it on or off) and therefore does not display a "toggle" but rather two "switches", one for turn on and the other for turn off. 
And if I use the website of wled to change the state it doesnt update in HomeAssistant.
This has to do with the "assumed_state" setting in HA, it can be overriden but wont change the fact that HA cannot query the light for its state.

I know this will all be fixed with the mqtt rewrite, but I wanted to ask if a state could be implementet in the mean time. 

Here is a small example with a picture and an explanation about this:
https://www.reddit.com/r/homeassistant/comments/c87o3h/why_is_the_toggle_button_on_these_two_lights/

**Comment by aidbish** [comment]
The post above says the status has just been implemented

**Comment by nightcat91** [comment]
> The post above says the status has just been implemented

Yes I saw that. 
Without having tried it yet, I think this only sends out a single "Online" message when the WLED connects to the MQTT broker and also sets up a LWT message with the broker that gets published should the WLED device go offline.

So this is only an online or offline state, but I was talking about the state beeing "on" or "off" in regards to the actual light beeing on or off.

**Comment by marc-gist** [comment]
@nightcat91 you can get "on/off" from the /g topic which is '0' for off, and >0 when on (master brightness setting if i remember correctly without checking docks :) )

**Comment by nightcat91** [comment]
Ok, I will have a look at that once I get home.
So far I was just using the Auto Discovery from HomeAssistant, so this would imply that the auto discovery message that gets send from WLED does not include this information about its state and beeing able to retrieve it from that topic

**Comment by marc-gist** [comment]
> Ok, I will have a look at that once I get home.
> So far I was just using the Auto Discovery from HomeAssistant, so this would imply that the auto discovery message that gets send from WLED does not include this information about its state and beeing able to retrieve it from that topic

Well, problem with HA is that it doesn't know to check that attribute for on/off, you will have to customize the code. so from that perspective, you are likely better to wait for Franck's new WLED integration in HA 102 release in a week or so (or upgrade to latest beta).

**Comment by nightcat91** [comment]
> > Ok, I will have a look at that once I get home.
> > So far I was just using the Auto Discovery from HomeAssistant, so this would imply that the auto discovery message that gets send from WLED does not include this information about its state and beeing able to retrieve it from that topic
> 
> Well, problem with HA is that it doesn't know to check that attribute for on/off, you will have to customize the entity in yaml. so from that perspective, you are likely better to wait for Franck's new WLED integration in HA 102 release in a week or so (or upgrade to latest beta).

Are you talking about this pr? 
https://github.com/home-assistant/home-assistant/pull/28542
Looks great, hope it works well with the next release.

And you are right, without messing around in the yaml entity or possible the mqtt light integration is hard to get it working as it should since there is no topic that provides "state" as documented for HA
https://www.home-assistant.io/integrations/light.mqtt/#command_topic

**Comment by Twsts** [comment]
Trying to control the white channel in my RGBW lights via Home Assistant - but can't figure it out. Anyone got it working yet? Via mqtt and "white_value_template" maybe?

Tried the WLED integration in the HA beta, works really well - but no white channel. Looking at the code, the RGBW support is mentioned tho.

**Comment by bodiroga** [comment]
Hi @timothybrown!

Your proposal looks great, but what do you think about adding a 'first-level-prefix' to all topics? What about 'wled'? For the discovery viewpoint, it's easier to subscribe to something like "wled/<device_id>/status", so that we can ensure that the device autodiscovered follows the wled topic convention. "<device_id>/status" seems too generic, in my opinion. This way, the final topics will be:

_Power Status Topic_: wled/<device_id>/switch => on|off
_Power Command Topic_: wled/<device_id>/switch/set <= on|true|1|off|false|0
_Color Status Topic_: wled/<device_id>/color => R,G,B
_Color Command Topic_: wled/<device_id>/color/set <= #HEX|R,G,B
_Brightness Status Topic_: wled/<device_id>/brightness => 0-255
_Brightness Command Topic_: wled/<device_id>/brightness/set <= 0-255
_Effect Status Topic_: wled/<device_id>/fx => FX Number
_Effect Command Topic_: wled/<device_id>/fx/set <= FX Number|FX Name
_API Response Topic_: wled/<device_id>/api => XML API Response
_API Command Topic_: wled/<device_id>/api/set <= HTTP API Format
_Availability Topic_: wled/<device_id>/status => online|offline

On the other hand, have you thought about marking some of the messages as retained?

Thanks for your attention and best regards,

Aitor

**Comment by c-soft** [comment]
Not putting any pressure, but how is this coming along? I'm trying to integrate RGBW strips with Homeassistant via MQTT and having issues with white. This change would remedy that issue :) Thanks!

**Comment by soundstorm** [comment]
@c-soft Hass has its own WLED Integration now and with 0.4.0 the hass auto discovery is disabled by default (you have to send an empty retained to that /config topic if you don't want a duplicate of your lights)

**Comment by c-soft** [comment]
Oh, seems like I'm quite behind :) Thanks, I will test it out!

**Comment by c-soft** [comment]
In case anybody else finds my comment above: I've managed to make it work now with official Home Assistant integration. Nice & easy, works like a charm. Thanks @soundstorm for pointing that out.

**Comment by weini22** [comment]
Stupid question: Has this already beed implmented in 0.9.1?
I tested e. g. Device Topic/Segment/fx => FX Number but didn´t get a proper response.

**Comment by shantur** [comment]
Is there any update on this? Cannot find any branch or fork

**Comment by DieNummerEins** [comment]
Watching!

**Comment by aidbish** [comment]
cant see it happening, now that HA has an official integration for wled that doesn't use MQTT

**Comment by shantur** [comment]
MQTT options being discussed here are much more powerful than whats provided in HA integration.
I am especially interested in Sections control.

**Comment by aidbish** [comment]
segments are being implemented in the integration. Being as the OP was looking at doing it for HA i doubt there will be anyone with HA now wanting to take it on.

**Comment by grumpazoid** [comment]
> 
> I'm interested to hear from anyone using the MQTT support in a non-HA setting (i.e., with custom code, other automation platforms and so forth). Would these topics be sufficient for your use? Is there anything else you'd like to see?
> 

Hi.
Firstly I've only just discovered WLED. Thanks for a great project.
My home automation consists of Domoticz and NodeRed which works great for me. It is a very flexible setup. MQTT messages are passed from Domoticz into Node Red and vica versa. 

Up untill now I have been using Tasmota on ESP8266. While this only offers basic support for LED's , by emulating a Hue Bridge, it does have native Domoticz support. All that is needed is to create a virtual device in domoticz and then type the IDX number from the virtual device into Tasmota.

So after playing with all the nice effects in WLED I jumped straight in to see how it works with MQTT which has led me here.

I currently have subscribed to wled/1/c and wled/1/g.
The first one gives me the colours as expected. /g just gives me an ouptut of 1 each time?

Thanks for all the hard work.
Roopz

**Comment by Aircoookie** [comment]
@grumpazoid  Welcome!
Glad you just got it working :)
If you are interested, WLED emulates parts of the Hue bridge as well for Alexa support. You can get the light id from the json reply to `[WLED-IP]/api/lights`

**Comment by grumpazoid** [comment]
@Aircoookie Thanks for that info. I have lots of things to play with now. Also https://github.com/frustreermeneer/domoticz-wled-plugin  which I am sure you have seen.

**Comment by ShuffleBit** [comment]
My $1.80 (2 cents, inflation adjusted) worth of MQTT remarks:

I control most things via in-house MQTT (Mosquitto). I'm no MQTT expert, but neither am I a casual user. MQTT allows me to perform complex monitoring and control without having to maintain a list of device/ipaddr pairs, and without the staggering bloat of HA or OpenHab. I have hundreds of devices scattered over several square miles. I can turn 75 disparate devices on with a single publish. A motion event a mile away can turn on driveway lighting if the ambient light is low, and trigger cameras to take a picture so TensorFlow can read/log the license plate. All of the things that happen in that sequence are contained in one MQTT payload, easily processed.

IMO, Tasmota does MQTT m/l perfectly: two configurable topics, and two configurable group topics to which it publishes all changes (plus occasional state messages), and to which it subscribes at startup for command requests. All of the MQTT payloads are based on the same set of commands that can be entered in Tasmota's web-ui console, MQTT payload, or the HTTP api, all 100% consistent.

I'm not saying WLED should implement the same thing, I'm only saying I really like the way Tasmota does MQTT, and I believe that Tasmota is a great example of getting it right.

**Comment by hazymat** [comment]
I'm not sure where we are with the MQTT restructuring (please forgive me that I didn't have the stamina to read all comments on this thread!)

But I would like to concur with the above comment from @ShuffleBit to say that the Tasmota project has done an excellent job in thinking deeply about MQTT structures, and although I don't think it's perfect, it would be silly to try to re-invent the wheel on this one.

The other thing I'd add is that I don't think it's a good idea to build an MQTT structure primarily with a particular automation package in mind. Yes I know HomeAssistant seems to be the most popular right now and hopefully it will have a long future. But I think WLED should think a bit bigger than that, and push forward with its own structure if that's the right thing to do.

**Comment by JAMESBOWLER** [comment]
If there is 5 ways to do something you can have a standard to cover them all then you have 6 ways to do something.

If you look into some of the option's out there the https://homieiot.github.io/specification/ is a pretty mature solution. 
If wled were to align their MQTT to a current standard then no new one need to be created. You may need to have something added to the current standard. Then openHAB and HA would just discover the device and add it to the list.

**Comment by hazymat** [comment]
To answer the specific question in opening post by @timothybrown _"I'm interested to hear from anyone using the MQTT support in a non-HA setting (i.e., with custom code, other automation platforms and so forth). Would these topics be sufficient for your use? Is there anything else you'd like to see?"_ here is my answer:

I am using the MQTT support in NodeRED and have built my own **NodeRED discovery platform for multiple device types** (Shelly, Sonoffs running Tasmota, arduinos running custom code).

It queries these devices over MQTT to provide an interface showing customised details. See screenshot below.

I would love to include my WLEDs in here, however they would at the very least need to provide IP address over MQTT, for example when a query is sent over MQTT to the group topic.

![screenshot-20201103-161340](https://user-images.githubusercontent.com/7063284/98011549-ea496580-1def-11eb-9258-ed6dc0d5cc48.png)

**Comment by janbernhardt** [comment]
Hi, 
thanks for this awesome project! I love it!
Regarding refactoring the MQTT topics, I would like to request a "button" state topic.
I think this would be very useful when connecting with home automations systems to detect and react on button events.
My idea would be to connect a motion detection sensor as my WLED button. However depending on the time of the day I would like to run different effects on my WLED controller. So the idea would be to subscribe homeassistant with the button topic of WLED and on "button/motion" detection homeassistant would derive the correct effect and brightness to be used and sends back a command to the matching */set topics to run WLED light effects.

**Comment by smalljoshua** [comment]
Is there any current documentation on the MQTT implementation as it stands today? I'm waiting on hardware to arrive but want to start getting something sketched out.

Obviously, the Wiki still reflects the old set of topics.

I'd ideally like to be able to remotely read and set the various parameters of effect/speed/colour etc from a control system and store these to presets plus recall these presets.

Whilst I could use the HTTP API, I'd like to give MQTT a go first.

**Comment by hazymat** [comment]
I believe the existing MQTT docs are still valid. I'm running 0.11.0 and this still works:

![image](https://user-images.githubusercontent.com/7063284/102463422-fbb69c00-4042-11eb-9574-f590261f1b5b.png)

**Comment by smalljoshua** [comment]
> (Note, API is used through MQTT, still better than HTTP in my opinion as you don't need IP addresses of WLED controllers)

That's a very fair point, I'm working on integrating with Crestron Control so no matter what happens I need to play with strings anyway.

**Comment by nodiaque** [comment]
while this is nice, you only control the main color channel using this. Using http json, you get control over all 3 channel of each segment

**Comment by RDAppel** [comment]
Any thoughts about using a command/status prefix like they do with Tasmota? They use "cmnd" and "stat." For ex: cmnd/wled/color would change the color, while stat/wled/color would be the status. This is also just the default (%prefix%/%topic%), and can be changed by the user. So, if you wanted you could flip it (%topic%/%prefix%) and get wled/color/cmnd.

I use a lot of Tasmota and WLED devices (I'm assuming a lot of other people do too), and it would be really convenient if these protocols matched!

**Comment by RDAppel** [comment]
> My $1.80 (2 cents, inflation adjusted) worth of MQTT remarks:
> 
> I control most things via in-house MQTT (Mosquitto). I'm no MQTT expert, but neither am I a casual user. MQTT allows me to perform complex monitoring and control without having to maintain a list of device/ipaddr pairs, and without the staggering bloat of HA or OpenHab. I have hundreds of devices scattered over several square miles. I can turn 75 disparate devices on with a single publish. A motion event a mile away can turn on driveway lighting if the ambient light is low, and trigger cameras to take a picture so TensorFlow can read/log the license plate. All of the things that happen in that sequence are contained in one MQTT payload, easily processed. "Master, Paul's pickup truck will arrive in two minutes."
> 
> IMO, Tasmota does MQTT m/l perfectly: two configurable topics, and two configurable group topics to which it publishes all changes (plus occasional state messages), and to which it subscribes at startup for command requests. All of the MQTT payloads are based on the same set of commands that can be entered in Tasmota's web-ui console, MQTT payload, or the HTTP api, all 100% consistent.
> 
> I'm not saying WLED should implement the same thing, I'm only saying I really like the way Tasmota does MQTT, and I believe that Tasmota is a great example of getting it right.

This has my vote (and my $.02). I love how Tasmota implemented this.

**Comment by ChessSpider** [comment]
Dont know if anyone cares, but https://github.com/Aircoookie/WLED/pull/1588/files uses automatic device discovery in home assistant.  Feedback welcome, saw this discussion too late anyway,

**Comment by loganbest** [comment]
I understand this may be a backburner task or maybe even no longer considered a task since the native hass integration has happened since this issue creation, but I and many others still think there's a lot of value in making WLED's MQTT implementation at least compatible with HA's auto discovery to start out. Can we get a definitive answer on where the status of this is and if it's in any future road maps for the project?

**Comment by hazymat** [comment]
I would agree with the above, and would also like to add that HASS is by no means the only system used by folk for home automation. MQTT is still the protocol of choice for agnostic support, i.e. across different systems.

As someone who uses MQTT religiously (not just for control, but for device monitoring and data collection), it's something I really want to see.

One of the main difficulties I've had is being able to control individual pixels completely over MQTT; this requires defining and querying segments on the fly. Currently I have implemented my own MQTT protocol for my little devices around the home running my own sketches, but I'd like to migrate them to WLED once there is support for segment management over MQTT. 

Sure I know this sounds like a niche case, perhaps it is - but it's something I would very much appreciate being able to do, and it's perhaps one of hundreds of niche things that could be done if a more comprehensive support for MQTT were available.

**Comment by blazoncek** [comment]
It is possible to send anything to WLED using JSON API over MQTT.
Just make sure your MQTT message doesn't exceed 1024 bytes (which can be rather quickly).
If you depend on larger messages increase MQTT buffer size in platformio.ini

**Comment by SirGoodenough** [comment]
Being able to OTA update all 35 of my Tasmota devices with 1 MQTT command is something that I would like to see in the wLED a MQTT stack.  A group topic that they all have in common, and additionally a couple of group topics that can be set to make smaller groups.  Plus an update method to latest.
I'm not looking to full control of the devices, I'm looking for things like on, off, OTA update, and basic maintenance things thru MQTT.
The Tasmota style interface is overkill for me for this, but is an excellent model IMHO.

Shorter term, a little more Documentation on what is what now would be helpful.  
I only see docs dating back to version 8 so not sure if they are valid.

**Comment by blazoncek** [comment]
MQTT in WLED may not get such love.
For help with updating multiple devices there is a shell/command prompt script in tools subfolder.
You can download and modify it for your needs.

As for documentation you are welcome to submit PR and make it better. Use [WLED docs](https://github.com/Aircoookie/WLED-docs).

**Comment by WhiteLionATX** [comment]
can someone tell my how to set the secondary and third color by MQTT. I tried to send msg.payload to "deviceTopic/api" like that:
`//msg.payload = {"col":[[255,0,0], [0,255,0], [0,0,255]]}; // {"seg":{"col":[[255,0,0], [0,255,0], [0,0,255]]}};
//msg.payload = {"seg":[{"col":[null,null,[0,0,247]]}]};
//msg.payload = {"seg":{"0":[0,8,[255,0,0], 10,18,[0,0,255]]}};
msg.payload = {"seg":[{"col":[["#00FFFF"]]}]};
return msg;`
But all failed .... Primary works like charm: "deviceTopic/col" msg.payoad "#00ff00"
Thx!

**Comment by satellite-no** [comment]
So WLED Docs link to this issue but seems as though it has _died_.  Lots of discussions and comments but doesn't seem like the originator is not working on this anymore.  Should the docs be updated to no longer reference this since it's sort of deceiving now (giving the impression a change is coming)?   

Doc link https://kno.wled.ge/interfaces/mqtt/

**Comment by blazoncek** [comment]
@satellite-no What is the issue here? Link to the stale GH issue or MQTT implementation not adhering to docs?

If it is the former, please fork WLED docs repo, make change and do a PR.
If it is the latter, pinpoint discrepancy and open a new issue here at WLED repo.

It has been quite a while since anybody touched MQTT.

**Comment by mylesdebastion** [comment]
Building an escape room that uses MQTT to update the status of various clients (puzzles) on the network including WLED. Would like to trigger different effect patterns using MQTT, is this actually possible in WLED without diving into the source code?

**Comment by blazoncek** [comment]
MQTT allows for full JSON API control.
Use `/api` subtopic.

**Comment by brackendegn** [comment]
Curious as to what the status of this is? Has there been any progress in being able to get a JSON response back as opposed to XML? I am working on a custom app that would allow for "cloud" access to  WLED controllers. Being able to get a state response back in JSON would be very useful for this. I am happy to help in any way I can, though my skills in that area are not very good.

**Comment by blazoncek** [comment]
FYI JSON can get quickly over 10kB while MQTT buffers are usually 1kB.

**Comment by brackendegn** [comment]
I have configured both my mqtt broker and WLED to use a buffer of 3KB and have it sending JSON to the /v topic instead of XML, so far it seems to be working fine

**Comment by blazoncek** [comment]
> buffer of 3KB

Use ESP32 and configure 32 segments and use 0_15 branch. Then try again.
JSON size is almost proportional to number of segments.

**Comment by brackendegn** [comment]
I used 3KB as the buffer size, as it was plenty for my needs. I just tested a 10KB buffer and I can get up to 18 segments, but after that it stops sending the state updates through MQTT. Seems to be possibly another limiter in WLED somewhere, but I can't figure out where. The difference in state JSON from 18 to 19 segments is only around 200 Bytes(5.85KB vs. 5.65KB). There seems to be a limiter around 5.7KB.

18 Segments is more than enough for my use case. If this isn't a need for others, I will just leave it as is.

**Comment by MikeEitel** [comment]
Hi
Is there still some work done on the mqtt subject?
And did anyone got the request for a PRESET working?
Tried different ways to write in the /api  but none worked also I have `all the usual working fine.`
I 


P.S.
Here a small addon i included in the publish function to get some wifi additional data as I could not get the INFO running,

char subuf2[48];
// Create a standalone XML structure for Wi-Fi data
DynamicBuffer wifiBuf(256); // Allocate a buffer for Wi-Fi XML
bufferPrint wifiPbuf(wifiBuf.data(), wifiBuf.size());

// Build the Wi-Fi XML structure
wifiPbuf.print(F("<?xml version=\"1.0\" ?>"));
wifiPbuf.print(F("<wifi>"));
wifiPbuf.printf_P(PSTR("<ssid>%s</ssid>"), WiFi.SSID().c_str());
wifiPbuf.printf_P(PSTR("<rssi>%d</rssi>"), WiFi.RSSI());
wifiPbuf.printf_P(PSTR("<channel>%u</channel>"), WiFi.channel());
wifiPbuf.print(F("</wifi>"));

// Publish the Wi-Fi XML to a separate MQTT topic
strlcpy(subuf2, mqttDeviceTopic, 33);
strcat_P(subuf2, PSTR("/wifi"));
mqtt->publish(subuf2, 0, retainMqttMsg, wifiBuf.data(), wifiPbuf.size());

**Comment by artemzhilin82** [comment]
[МайкЭйтель](https://github.com/MikeEitel) I did this for the first PRESET(wled/sound/api ={"ps":"1"}

**Comment by rlimberger** [comment]
> I have configured both my mqtt broker and WLED to use a buffer of 3KB and have it sending JSON to the /v topic instead of XML, so far it seems to be working fine

how did you configure /v to respond in JSON?

**Comment by blazoncek** [comment]
It will not. It only accepts JSON.

**Comment by brackendegn** [comment]
It’s possible. I have it working, but if you go over a certain amount of segments it won’t work. I’ve managed to get it up to 18 segments with no problems. You’ll have to do it on your own version though. I don’t think it’s going to be in the official WLED.

**Comment by sanderkooger** [comment]
Would
```
Color Command Topic
Device Topic/color/set <= #HEX|R,G,B 
```

Not create an issue for people running RGBW strips. It might be a good idea to think of a method to include white levels.

**Comment by WouterGritter** [comment]
I had some thoughts on this in #4305 which I have closed. I'm not in need of this change anymore, but it might be worth looking at that issue for a possible different use-case of MQTT topics.
